lib/std/debug.zig

```diff
@@ -50,163 +40,378 @@ pub const LineInfo = struct {
+    /// Otherwise, this will would call panicTerminate recursively and deadlock.
+    pub const panicTerminate = os.abort;
+};
+
+const is_default_config = @hasDecl(root, "debug_config");
+const config = if (is_default_config)
+    root.debug_config
+else
+    default_config;
+
+const has_writer_decl = @hasDecl(config, "getWriter");
+
+/// The function used to get a writer for debugging.
+/// Slightly different name than in config to avoid redefinition in default.
+const getWrit = if (has_writer_decl)
+    config.getWriter
+else
+    default_config.getWriter;
+
+/// The function used to determine which escape codes can be used.
+/// Slightly different name than in config to avoid redefinition in default.
+const getTTYConfig = if (@hasDecl(config, "detectTTYConfig"))
+    config.detectTTYConfig
+else if (has_writer_decl)
+    @compileError("getWriter exists in config, so detectTTYConfig must also exist")
+else
+    default_config.detectTTYConfig;
+
+const has_address_symbol_mapping_decl = @hasDecl(config, "AddressSymbolMapping");
+const using_default_address_symbol_mapping = is_default_config or
+    !has_address_symbol_mapping_decl;
+
+/// Slightly different name than in config to avoid redefinition in default.
+const AddrSymMapping = if (has_address_symbol_mapping_decl)
+    config.AddressSymbolMapping
+else
+    default_config.AddressSymbolMapping;
+
+/// Slightly different name than in config to avoid redefinition in default.
+const writeLineFromSourceFile = if (@hasDecl(config, "attemptWriteLineFromSourceFile"))
+    config.attemptWriteLineFromSourceFile
+else
+    default_config.attemptWriteLineFromSourceFile;
+
+/// Slightly different name than in config to avoid redefinition in default.
+const panicTerm = if (@hasDecl(config, "panicTerminate"))
+    config.panicTerminate
+else
+    default_config.panicTerminate;
+
+var mutex = std.Thread.Mutex{};
+
+const HeldWriter = HeldValue(@TypeOf(getWrit()));
+pub fn heldWriter() HeldWriter {
+    return .{ .value = getWrit(), .held = mutex.acquire() };
+}
 
 /// Deprecated. Use `std.log` functions for logging or `std.debug.print` for
 /// "printf debugging".
 pub const warn = print;
 
-/// Print to stderr, unbuffered, and silently returning on failure. Intended
+/// Print to log writer, unbuffered, and silently returning on failure. Intended
 /// for use in "printf debugging." Use `std.log` functions for proper logging.
 pub fn print(comptime fmt: []const u8, args: anytype) void {
-    const held = stderr_mutex.acquire();
+    const held = heldWriter();
     defer held.release();
-    const stderr = io.getStdErr().writer();
-    nosuspend stderr.print(fmt, args) catch return;
-}
-
-pub fn getStderrMutex() *std.Thread.Mutex {
-    return &stderr_mutex;
+    const writer = held.value;
+    nosuspend writer.print(fmt, args) catch {};
 }
 
 /// TODO multithreaded awareness
-var self_debug_info: ?DebugInfo = null;
+var symbol_at_address: ?AddrSymMapping = null;
```

Comment:
While we're at it let's consider if this global cache should stay, the hard part is deciding when to deinitialize it.